### ISSUE TITLE
Chest furnature causes glitches and crashing in some cases.

### ISSUE THREAD

**Issue by CatOfBlades** [konchunas/minicraft-psp#5]
The Chest object was tested in a ps-vita psp emulator and infrequently causes crashes.
Upon analysis of the source, a likely cause of this might be the inventory implementation.

Potential problem code in ./entity/Inventory.cpp:104 :
<code>
Item* Inventory::removeAt(int index)
{
	deque<Item*>::iterator it = items.begin() + index; //Do not offset from an iterator in a deque
	Item * item = *it;	//remember item beforehand
	items.erase(it);	//remove item from list
	return item;		//return item
}
</code>

Possible replacement:
<code>
Item* Inventory::removeAt(int index)
{
        deque<Item*> buffer; //temporary storage space
        while(index > 0) //move items out of the way
        {
                index--;
                buffer.push_front(items.front()); //move items to storage
                items.pop_front(); //remove items from inventory
        }
        Item * item = items.front(); // get item at iterator
        items.pop_front(); // erase item
        while(buffer.size > 0) //put items back
        {
                items.push_front(buffer.front()); //move items from storage
                buffer.pop_front(); //remove items from storage
        }
	return item;		//return item
}
</code>

**Comment by konchunas** [comment]
Thank you for interest.
Have you tested this solution?

Anyway, what kind of glitches do you get? Something with first and last items? Maybe some random mix ups?

**Comment by CatOfBlades** [comment]
I just created a branch which seems to fix this glitch.
I am not sure that this code is the one causing the problem, however the data structure used in this case is not handled properly. Deque's are open ended que-stacks, not general purpose inventories.
"accessing elements in a deque by offsetting a pointer to another element causes undefined behavior." Quote from: http://www.cplusplus.com/reference/deque/deque/

**Comment by konchunas** [comment]
The thing is It does not change the pointer itself. That code adds index to deque iterator and for random access iterator it seems to be totaly okay.
http://www.cplusplus.com/reference/iterator/RandomAccessIterator/

The only flaw I see is not checking if resulting iterator equals to end()

Anyways I want to try to repeat this bug on real hardware. Do you have some steps so I can see the bug myself?

**Comment by CatOfBlades** [comment]
The bug occurs randomly and I have not tested it on a real PSP. The instances in witch it occurred were on a PSVita and happened when accessing the inventory of a chest, or when adding or removing an item from a chest. The errors which have occurred include vanishing items and game crashing. I will do more play testing to see if I can reliably reproduce the bug on an emulator.

**Comment by CatOfBlades** [comment]
It seems you were right, I jumped to conclusions. The code IS completely valid. I just figured that a que-stack would fall over if you pulled something out of the middle. But they do in fact seem to overload the arithmetic operators for random access iterators for each class. The only point left to argue is that erasing something in the middle of a deque invalidates all pointers previously made to it. As there seems to be no pointer misuse in the project, I can only conclude that I have lernt something.
I will keep looking to see if I can find the problem. PS: Is there any reason that we should not use a list or a vector for the inventory?

**Comment by CatOfBlades** [comment]
I have pinpointed a list of actions which reliably reproduce the bug and if overdone crashes the system. I have tested this on a psvita, psp, and the ppsspp emulator. My modification to the code had no effect on the glitch, it seems to happen in both versions. To reproduce the glitch, first put more then one of a stack-able item in a chest, second get more of that item and wield it and open the chest, third withdraw the items from the chest, which are turned into a duplicate of the bottom item of your inventory.

**Comment by konchunas** [comment]
I probably have found the reason of glitches and crashes while working with chest. You can download binary version with fix here: https://github.com/konchunas/minicraft-psp/blob/master/build/minicraft_1.0.1b_unsigned.zip

But, the problem you described with unweilding stuff is not related to chest management problem.
Problem was in trying to put item from chest into undefined inventory slot and vice versa.
But I also fixed bug with resource entry duplication.

Could you please confirm fixes work for you?

**Comment by CatOfBlades** [comment]
Yes that does fix it. Though it seems odd to not have it add to the ones in your hand. Could it be modified to treat the hand as an extra inventory space and move the item there if it is of the same type?
